### PR TITLE
refactor(abigen): derives, struct expansion

### PR DESCRIPTION
**Pull request by DaniPopes**
## Solution

Minor refactors that didn't make the cut for their own PR

- add more trait implementations and make existing ones more explicit (`impl Clone` -> `impl ::core::clone::Clone` etc)
- unify `expand_data_{struct,tuple}` to use a single generic function
- rename `event_derives` to `extra_derives` and standardize how `can_derive_default` is added to the final item derives
- replace empty `TokenStream` with `Option<TokenStream>`

## PR Checklist

-   [ ] Added Tests
-   [ ] Added Documentation
-   [ ] Updated the changelog
-   [ ] Breaking changes

### CODE REVIEW

File: ethers-contract/ethers-contract-abigen/src/contract.rs [hunk 1]
```diff
@@ -52,14 +51,19 @@ impl ExpandedContract {
             abi_structs,
             errors,
         } = self;
+
         quote! {
-           // export all the created data types
             pub use #module::*;
 
+            /// This module was auto-generated with ethers-rs Abigen.
+            /// More information at: <https://github.com/gakonst/ethers-rs>
```
Comment (mattsse) [inline ethers-contract/ethers-contract-abigen/src/contract.rs:59]: great!

File: ethers-contract/ethers-contract-abigen/src/contract.rs [hunk 2]
```diff
@@ -103,7 +107,7 @@ pub struct Context {
     error_aliases: BTreeMap<String, Ident>,
 
     /// Derives added to event structs and enums.
-    event_derives: Vec<Path>,
+    extra_derives: Vec<Path>,
```
Comment (mattsse) [inline ethers-contract/ethers-contract-abigen/src/contract.rs:110]: this is so much better :)

**Review by mattsse — APPROVED** [review]
this is so much better.

tysm for showing this some love!

File: ethers-contract/ethers-contract-abigen/src/contract/structs.rs [hunk 3]
```diff
@@ -592,16 +582,56 @@ fn struct_type_projections(name: &str) -> Vec<String> {
     iter.rev().map(str::to_string).collect()
 }
 
+pub(crate) fn expand_struct(
```
Comment (gakonst) [inline ethers-contract/ethers-contract-abigen/src/contract/structs.rs:585]: clean!

File: ethers-contract/ethers-contract-abigen/src/contract.rs [hunk 4]
```diff
@@ -322,6 +321,101 @@ impl Context {
...
+
+        quote! {
+            // The `Lazy` ABI
+            #abi
+
+            // The static Bytecode, if present
+            #bytecode
+
+            // Struct declaration
+            pub struct #name<M>(#ethers_contract::Contract<M>);
```
Comment (gakonst) [inline ethers-contract/ethers-contract-abigen/src/contract.rs:385]: we'll need to resolve conflict with https://github.com/gakonst/ethers-rs/commit/9733b16afc5c35be4409b72c172d8f7405f45d7c

**Review by gakonst — CHANGES_REQUESTED** [review]
nice work!

File: ethers-contract/ethers-contract-abigen/src/util.rs [hunk 5]
```diff
@@ -3,75 +3,85 @@ use eyre::Result;
...
+///   preserve_underscore_delim("pascalCase", "__pascalcase__"),
+///   "__pascalCase__"
+/// );
+/// ```
+pub(crate) fn preserve_underscore_delim(ident: &str, original: &str) -> String {
+    let is_underscore = |c: &char| *c == '_';
+    let pre = original.chars().take_while(is_underscore);
+    let post = original.chars().rev().take_while(is_underscore);
+    pre.chain(ident.chars()).chain(post).collect()
 }
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply ethers-contract/ethers-contract-abigen/src/util.rs:78]: Rewrote it to make it more idiomatic using more variables, don't think there's another way to do it

**Comment by gakonst** [comment]
ty!

**Pull request merged**